wasi/wasi_test.go

```diff
@@ -0,0 +1,261 @@
+		},
+		{
+			name: "utf-8 string",
+			// "😨", "🤣", and "️🏃‍♀️" have 4, 4, and 13 bytes respectively
+			args: []string{"😨🤣🏃\u200d♀️", "foo", "bar"},
+			expectedArgs: [][]byte{
+				[]byte("😨🤣🏃\u200d♀️\x00"),
+				[]byte("foo\x00"),
+				[]byte("bar\x00"),
+			},
+			expectedBufSize: 30,
+		},
+		{
+			name: "invalid utf-8 string",
+			args: []string{"\xff\xfe\xfd", "foo", "bar"},
+			expectedArgs: [][]byte{
+				[]byte("\xff\xfe\xfd\x00"),
+				[]byte("foo\x00"),
+				[]byte("bar\x00"),
+			},
+			expectedBufSize: 12,
+		},
+	}
+
+	for _, tt := range tests {
+		tc := tt
+
+		t.Run(tc.name, func(t *testing.T) {
+			opts := []Option{}
+			if tc.args != nil {
+				argsOpt, err := Args(tc.args)
+				require.NoError(t, err)
+				opts = append(opts, argsOpt)
+			}
+			wasiEnv := NewEnvironment(opts...)
+			hostFunctionCallContext := buildMockHostFunctionCallContext()
+
+			// Serialize the expected result of args_size_get
+			argCountPtr := uint32(0) // arbitrary valid address
+			expectedArgCount := make([]byte, SIZE_UINT32)
+			binary.LittleEndian.PutUint32(expectedArgCount, uint32(len(tc.args)))
+			bufSizePtr := uint32(0x100) // arbitrary valid address that doesn't overwrap with argCountPtr
+			expectedBufSize := make([]byte, SIZE_UINT32)
+			binary.LittleEndian.PutUint32(expectedBufSize, tc.expectedBufSize)
+
+			// Compare them
+			errno := wasiEnv.args_sizes_get(hostFunctionCallContext, argCountPtr, bufSizePtr)
+			require.Equal(t, ESUCCESS, errno)
+			require.Equal(t, expectedArgCount, hostFunctionCallContext.Memory.Buffer[argCountPtr:argCountPtr+4])
+			require.Equal(t, expectedBufSize, hostFunctionCallContext.Memory.Buffer[bufSizePtr:bufSizePtr+4])
+
+			// Serialize the expected result of args_get
+			expectedArgs := make([]byte, SIZE_UINT32*len(tc.args))
+			argsPtr := uint32(0) // arbitrary valid address
+			expectedArgv := make([]byte, tc.expectedBufSize)
+			argvPtr := uint32(0x100) // arbitrary valid address that doesn't overwrap with argsPtr
+			argvWritten := uint32(0)
+			for i, arg := range tc.expectedArgs {
+				binary.LittleEndian.PutUint32(expectedArgs[argsPtr+uint32(i*SIZE_UINT32):], argvPtr+argvWritten)
+				copy(expectedArgv[argvWritten:], arg)
+				argvWritten += uint32(len(arg))
+			}
+
+			// Compare them
+			errno = wasiEnv.args_get(hostFunctionCallContext, argsPtr, argvPtr)
+			require.Equal(t, ESUCCESS, errno)
+			require.Equal(t, expectedArgs, hostFunctionCallContext.Memory.Buffer[argsPtr:argsPtr+uint32(len(expectedArgs))])
+			require.Equal(t, expectedArgv, hostFunctionCallContext.Memory.Buffer[argvPtr:argvPtr+uint32(len(expectedArgv))])
+		})
+	}
+}
+
+func TestArgsSizesGetReturnError(t *testing.T) {
+	dummyArgs := []string{"foo", "bar", "baz"}
+	argsOpt, err := Args(dummyArgs)
+	require.NoError(t, err)
+	wasiEnv := NewEnvironment(argsOpt)
+	hostFunctionCallContext := buildMockHostFunctionCallContext()
+
+	memorySize := uint32(len(hostFunctionCallContext.Memory.Buffer))
+	inMemory := uint32(0)
```

Comment:
document?

Reply:
Improved the comment. I though you mentioned about comments because it seems the even recent test codes don't seem to have godocs for each test functions, but let me know if your intention by `document` is not the comments.